Implements the proposed modifications from /u/akeniscool, with static methods to create immutable objects.

naroga opened the pull request:
As proposed by /u/akeniscool:

```
I think the flow and the resulting API is more complicated than it needs to be, blending together the parsing/factory portion of the library with the resulting UserAgent entity. I would rather see something like a static factory that is explicit about what data is being parsed:
$userAgent = UserAgentParser::parse($_SERVER['HTTP_USER_AGENT']);

// Equivalent default, yet still explicit.
$userAgent = UserAgentParser::parseFromGlobal();
These methods would parse the information specified, and return an immutable UserAgent entity (no setters!).
```

This PR removes all public setters from the UserAgent, makes the default constructor private and adds two methods to create immutable objects: `::create($string, $parser)` and `::createFromGlobal()`.

naroga committed:
Implements the proposed modifications from /u/akeniscool, with static methods to create immutable objects.

naroga committed:
Fixed version reference in README.md.

naroga commented:
Since there are no stable tags, I've also changed the documentation to install "dev-master" instead of "1.0.*". You are ready to mark this package as stable, add a tag (`git tag -a 1.0.0`, `git push --tags`) and change the README.md.

naroga committed:
Removes the  call so scrutinizer will stop complaining about it.

naroga committed:
Added 'Links' to the list of known real browsers (UNIX text-mode browser)

cryode commented:
/u/akeniscool here. While a step in the right direction, the `UserAgent` class is still doing too much, and you've only complicated its API further by adding more methods.

`UserAgent` should be the *result* of the parsing process, not the instigator. It has no idea how it was created, only that it holds the values of a parsed user agent string. The class itself holds little to no logic. The `UserAgentParser` class holds the parsing logic, and is responsible for generating `UserAgent` classes as results from parsing.

A more clear separation of concerns needs to happen.

Kudos to you for contributing to an open source package, though! :beer:

naroga commented:
> A more clear separation of concerns needs to happen.

That's a different PR, I guess. I've removed almost all public methods, and, even though there are two new methods (the factories), the public API is much cleaner now (I've disabled the public constructor, removed all setters). I'm not going to mess with the internals -- if @secondtruth wants to, he can do it without breaking BC, as the only public methods on `UserAgent` now are the factory methods. This PR only cares about the public API and the component usage.

secondtruth commented:
@naroga Thank you for contributing!

I prepared the `UserAgentStringParser` in the meantime. Let's see where it goes from here.

@cryode What do you think now?

cryode commented:
I personally wouldn't split up the Definition portion of the parser - I would just include it all as one class. Perhaps there's a benefit to breaking it out that I'm not aware of, but since there isn't really any complexity to this process, it just seems like an unnecessary step.

I also don't like that the parser belongs to the `UserAgent` class. Here is a very truncated example of how I would expose your lib's API:

```php
class UserAgentParser
{
	protected $knownBrowsers = [
		// ...
	];

	protected $knownBots = [
		// ...
	];

	public function parse($string, $strict = true)
	{
		// Do parsing here.

		return new UserAgent(
			$attributes['string'],
			$attributes['browserName'],
			$attributes['browserVersion'],
			$attributes['browserEngine'],
			$attributes['isKnownBrowser'],
			$attributes['operatingSystem'],
			$attributes['device'],
			$attributes['isBot']
		);
	}

	public function parseFromGlobal($strict = true)
	{
		$string = isset($_SERVER['HTTP_USER_AGENT']) ? $_SERVER['HTTP_USER_AGENT'] : null;

		return $this->parse($string, $strict);
	}

	public function addKnownBrowser($browser)
	{
		if (!in_array($browser, $this->knownBrowsers)) {
			$this->knownBrowsers[] = $browser;
		}
	}

	// Other definition getters, adders ...
}

class UserAgent
{
	protected $original;
	protected $browserName;
	protected $browserVersion;
	protected $browserEngine;
	protected $isKnownBrowser;
	protected $operatingSystem;
	protected $device;
	protected $isBot;

	public function __construct($original, $browserName, $browserVersion, $browserEngine, $isKnownBrowser, $operatingSystem, $device, $isBot)
	{
		$this->original = $original;
		$this->browserName = $browserName;
		$this->browserVersion = $browserVersion;
		$this->browserEngine = $browserEngine;
		$this->isKnownBrowser = $isKnownBrowser;
		$this->operatingSystem = $operatingSystem;
		$this->device = $device;
		$this->isBot = $isBot;
	}

	// Getters for attributes ...

	// Other methods like isBot() ...
}
```

You should only need to use the Parser as the public API. The `UserAgent` class, while technically available to instantiate, is effectively a value object for the parsed data. It is only a result.

cryode commented:
Example usage:

```php
$parser = new UserAgentParser();

$userAgent = $parser->parseFromGlobal();

var_dump($userAgent);

/*
object(UserAgent)[1]
  protected 'original' => string 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/45.0.2454.85 Safari/537.36' (length=120)
  protected 'browserName' => string 'Google Chrome' (length=13)
  protected 'browserVersion' => string '45.0.2454.85' (length=12)
  protected 'browserEngine' => string 'AppleWebKit' (length=11)
  protected 'isKnownBrowser' => boolean true
  protected 'operatingSystem' => string 'Mac OS X 10.10.5' (length=16)
  protected 'device' => string 'Macintosh' (length=9)
  protected 'isBot' => boolean false
 */
```

naroga commented:
@cryode `$knownBrowsers` and `$knownBots` could be a static public property, right?
Also, from your snippet:

    $parser = new UserAgentParser();
    $userAgent = $parser->parseFromGlobal();

This would contradict the original suggestion to make static factories.

    $userAgent = UserAgentParser::createFromGlobal(); //Or parseFromGlobal with you prefer.

This would be better, I guess.

cryode commented:
They could be static properties, yes. It's really up to the developer whether or not you want it to be completely static or instantiated (or both for whatever reason).

> This would contradict the original suggestion to make static factories.

It was just a suggestion, same as the code snippet above. More or less an example of how to separate the concerns from the two classes. Whether or not the end implementation is static or not is not up to me. It's not my job to tell another developer how to build their library in minute detail, or I'd just do it myself.

secondtruth commented:
> I personally wouldn't split up the Definition portion of the parser - I would just include it all as one class. Perhaps there's a benefit to breaking it out that I'm not aware of, but since there isn't really any complexity to this process, it just seems like an unnecessary step.

I did this mainly because Scrutinizer was complaining so much about the class size. Maybe I shouldn't trust it that much.

> I also don't like that the parser belongs to the `UserAgent` class.

I agree with you. This should be changed.

One note about the definition lists: The order of the entries is important. I came up with this interim solution:

```php
class Definition
{
    protected $list = array();

    public function __construct(array $list)
    {
        $this->list = $list;
    }

    public function toArray()
    {
        return $this->list;
    }

    public function add($name, $regex)
    {
        $this->list[$name][] = (string) $regex;

        return true;
    }

    public function addBefore($before, $name, $regex)
    {
        $pos = $this->find($before);

        if ($pos === false) {
            return false;
        }

        return $this->insertAt($pos, $name, $regex);
    }

    public function addAfter($after, $name, $regex)
    {
        $pos = $this->find($after);

        if ($pos === false) {
            return false;
        }

        return $this->insertAt($pos + 1, $name, $regex);
    }

    protected function insertAt($pos, $name, $regex)
    {
        if (isset($this->list[$name])) {
            return false;
        }

        array_splice($this->list, $pos, 0, [$name => (string) $regex]);

        return true;
    }

    protected function find($name)
    {
        return array_search($name, array_keys($this->list));
    }
}
```

naroga commented:
> I did this mainly because Scrutinizer was complaining so much about the class size. Maybe I shouldn't trust it that much.

If you truly want to separate the definitions from your code to appease Scrutinizer (and it's not a bad idea), you should probably make them definition/configuration files. Maybe a XML or a YAML. Otherwise, I'd say leave them with the UserAgent.

naroga commented:
Should I close this PR?

secondtruth commented:
No, I will merge it later. Thank you!
